### ISSUE TITLE
ORB_GPU generates random features on each run

### ISSUE THREAD

**Issue by alan-wong** [Itseez/opencv#6362]
Apologies if the following is insufficient, hopefully it describes the problem adequately.

- OpenCV version: 2.4.11
- Host OS: Windows 7

### In which part of the OpenCV library you got the issue?
Examples:
- cuda, features2d
- object tracking

### Expected behaviour
feature detecting should generate the same features on each run
### Actual behaviour
random features are detected, when you run the CPU implementation it generates the same features on each run
### Additional description

I've observed that when running the ORB_GPU to detect features that it will generate random features on each run on the same image, this doesn't happen when you run the CPU implementation version. I don't see any changes to the OpenCV 3 source code for orb.cpp so it doesn't look like this has been resolved.

### Code example to reproduce the issue / Steps to reproduce the issue

```
#include "opencv2/core.hpp"
#include "opencv2/features2d/features2d.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/calib3d/calib3d.hpp"
#include "opencv2/gpu/gpu.hpp"

using namespace std;
using namespace cv;

int main( int argc, char** argv )
{
  if( argc != 3 )
  { readme(); return -1; }

  Mat img_1 = imread( argv[1], CV_LOAD_IMAGE_GRAYSCALE );
  GpuMat gpuImg, mask;
  gpuImg.upload(img_1);

  if( !img_1.data || !img_2.data )
  { std::cout<< " --(!) Error reading images " << std::endl; return -1; }

  //-- Step 1: Detect the keypoints using ORB_GPU Detector
  cv::gpu ORB_GPU detector;

  std::vector<KeyPoint> keypoints_1, keypoints_2;

  detector( gpuImg, mask, keypoints_1 );
  detector( gpuImg, mask, keypoints_2 );

  Mat img_keypoints_1; Mat img_keypoints_2;

  drawKeypoints( img_1, keypoints_1, img_keypoints_1, Scalar::all(-1), DrawMatchesFlags::DEFAULT );
  drawKeypoints( img_1, keypoints_2, img_keypoints_2, Scalar::all(-1), DrawMatchesFlags::DEFAULT );

  //-- Show detected (drawn) keypoints
  imshow("Keypoints 1", img_keypoints_1 );
  imshow("Keypoints 2", img_keypoints_2 );

  waitKey(0);

  return 0;
}

/**
 * @function readme
 */
void readme()
{ std::cout << " Usage: ./SURF_detector <img1> <img2>" << std::endl; }
```

**Comment by timbobbarnes** [comment]
Hey @alan-wong, I looked into this a little bit yesterday. On my system (Ubuntu 14.04, 2.4 branch) I can verify that the keypoints are different between runs on GPU but not CPU... but not that different, if that makes sense. Sorting by position and comparing the lists shows that >90% of them are the same. My only guess right now is that successive runs may be selecting different sets of 'worst' keypoints to throw away, but I'm not really confident that's it.

I ran (a corrected version of) your code on an [opencv gpu test image](https://github.com/Itseez/opencv_extra/blob/master/testdata/gpu/features2d/aloe.png). Is this similar to what you're seeing?

![aloe1](https://cloud.githubusercontent.com/assets/1680498/14588506/fd0ad8ae-047f-11e6-8a4b-5eb3b4f02779.png) ![aloe2](https://cloud.githubusercontent.com/assets/1680498/14588508/008b98ba-0480-11e6-9d46-e1ccc6ef3772.png)

If so, I'll keep looking into it.

Here's the code I used:

```c++
#include <algorithm>
#include <iostream>

#include <opencv2/opencv.hpp>
#include <opencv2/gpu/gpu.hpp>

using namespace std;
using namespace cv;
using namespace cv::gpu;

void readme();

bool position_compare(const KeyPoint& a, const KeyPoint& b)
{
    return (b.pt.y > a.pt.y) || ((b.pt.y == a.pt.y) && (b.pt.x > a.pt.x));   
}

int main( int argc, char** argv )
{
    if( argc != 2 )
    { readme(); return -1; }

    Mat img_1 = imread( argv[1], IMREAD_GRAYSCALE );
    GpuMat gpuImg, mask;
    gpuImg.upload(img_1);

    if( !img_1.data )
    { cout<< " --(!) Error reading images " << endl; return -1; }

    //-- Step 1: Detect the keypoints using ORB_GPU Detector
    cv::gpu::ORB_GPU detector;

    std::vector<KeyPoint> keypoints_1, keypoints_2;

    detector( gpuImg, mask, keypoints_1 );
    detector( gpuImg, mask, keypoints_2 );

    if (keypoints_1.size() != keypoints_2.size())
    {
        cerr << "Keypoint vector lengths don't match" << endl;
        return 1;
    }
    std::sort(keypoints_1.begin(), keypoints_1.end(), position_compare);
    std::sort(keypoints_2.begin(), keypoints_2.end(), position_compare);
    for (std::size_t i = 0; i < keypoints_1.size(); i++)
        if (keypoints_1[i].pt != keypoints_2[i].pt)
        {
            cerr << "Keypoint " << i << " doesn't match " << keypoints_1[i].pt << ", " << keypoints_2[i].pt << endl;
        }
  

    Mat img_keypoints_1; Mat img_keypoints_2;

    drawKeypoints( img_1, keypoints_1, img_keypoints_1, Scalar::all(-1), DrawMatchesFlags::DEFAULT );
    drawKeypoints( img_1, keypoints_2, img_keypoints_2, Scalar::all(-1), DrawMatchesFlags::DEFAULT );

    //-- Show detected (drawn) keypoints
    imshow("Keypoints 1", img_keypoints_1 );
    imshow("Keypoints 2", img_keypoints_2 );
    imwrite("out1.png", img_keypoints_1);
    imwrite("out2.png", img_keypoints_2);

    waitKey(0);

    return 0;
}

/**
* @function readme
*/
void readme()
{ std::cout << " Usage: ./ORB_detector <img>" << std::endl; }
```

**Comment by alan-wong** [comment]
Thanks @timbobbarnes I'll have a look tomorrow on my dev machine, my laptop has an AMD card. The problem with this randomness is that it makes it hard to test whether my image processing and object tracking is improving the accuracy when using gpu ORB as it can generate a significantly different homography result on each run so I have to switch to CPU implementation for evaluation purposes, I have seen quite a few people mentioning the same thing: http://answers.opencv.org/question/58087/orb-gpu-random/ and I found someone else reported something about the FAST feature detector on the NVidia Forum which I can't find a link for at the moment. Alan.

**Comment by timbobbarnes** [comment]
OK, great, let me know after you get a chance to try it out. Thanks for the heads up about the other links, I'll take a look around and look at the FAST gpu code as well.

**Comment by alan-wong** [comment]
@timbobbarnes Hi, I ran against that test image and observed the following:
For 500 features I observed 421, 355, 353 and 317 different keypoints so this does appear to be very random. Also the NVidia cuda forum question I was referring to was more about the random performance times rather than the random features being generated: https://devtalk.nvidia.com/default/topic/767785/?comment=4284965

**Comment by timbobbarnes** [comment]
Hey @alan-wong, I've made some progress, but I'm not all the way there yet.

The first section of 'randomness' comes from the parallel nature of the FAST detector; you can see in [this section](https://github.com/Itseez/opencv/blob/f0b19d4659045b00c55f849187cd657b21a13e5d/modules/gpu/src/cuda/fast.cu#L268-L276) that the separate kernels add keypoints on a first-come-first-serve basis, so small variations in the execution time can give different responses. One can get around this by allocating way more memory than you probably need:

```c++
cv::gpu::FAST_GPU detector(20, true, 1); // '1' prohibits throwing away any keypoints
```

There's some other source, however, maybe something to do with `cv::gpu::ORB_GPU::cull()` that causes issues, I'll try and look into that over the next few days.

**Comment by timbobbarnes** [comment]
Hey @alan-wong, can you give this branch a try and let me know if it works for you?

https://github.com/timbobbarnes/opencv/tree/gpu-orb-repeatable-wip

On my machine, the last 5&ndash;20 keypoints come out in different orders on different runs, but otherwise they're the same. I'm not sure why that is, since they all get sorted by position and then response.

The only way I know to get this repeatability is to sort the keypoints, which takes away some of the speed advantage; I'm not sure whether this would be classified as a bug or feature after all&hellip;

**Comment by alan-wong** [comment]
@timbobbarnes thanks I'll test it out when I get back to work and let you know the results. I'm on paternity leave now so will be back in 2 weeks, thanks for looking into this

**Comment by timbobbarnes** [comment]
@alan-wong Hey, congratulations!! Enjoy your leave, get as much sleep as you can, talk to you in a few weeks.

**Comment by alan-wong** [comment]
@timbobbarnes Hi Tim, back in work and tried your suggested changes and they work! I tried on the test image and on our integrated application and can confirm that it is indeed generating the same features on every run, many thanks for your assistance.

**Comment by paroj** [comment]
I do not think this is a bug after all:

The ORB detector relies on filtering the image first by FAST and then by Harris responses.
As has been discussed the parallel FAST implementation is not deterministic.
The proposed solution effectively disables FAST filtering by computing it at every pixel.

Next the determinism is introduced by sorting the features by distance to the top left corner and keeping N. This is random. What we actually want is to sort the features by Harris response and only then keep N.

However at this point (feature at each pixel) we can just use Harris instead.

In conclusion ORB does not guarantee to find the N *best* features and the CUDA implementation just emphasizes on it by returning one of the possible solutions randomly while the CPU implementation always returns the same solution due to the sequential nature.
